File: packages/sdk-go/mailat/webhooks.go
package mailat

import (
	"context"
	"encoding/json"
	"fmt"
)

// WebhooksService handles webhook operations.
type WebhooksService struct {
	client *Client
}

// Create creates a new webhook endpoint.
func (s *WebhooksService) Create(ctx context.Context, req *CreateWebhookRequest) (*Webhook, error) {
	data, err := s.client.request(ctx, "POST", "/webhooks", req, nil)
	if err != nil {
		return nil, err
	}

	var resp Webhook
	if err := json.Unmarshal(data, &resp); err != nil {
		return nil, fmt.Errorf("failed to parse response: %w", err)
	}

	return &resp, nil
}

// Get retrieves a webhook by UUID.
func (s *WebhooksService) Get(ctx context.Context, webhookID string) (*Webhook, error) {
	data, err := s.client.request(ctx, "GET", "/webhooks/"+webhookID, nil, nil)
	if err != nil {
		return nil, err
	}

	var resp Webhook
	if err := json.Unmarshal(data, &resp); err != nil {
		return nil, fmt.Errorf("failed to parse response: %w", err)
	}

	return &resp, nil
}

// List retrieves all webhooks.
func (s *WebhooksService) List(ctx context.Context) ([]Webhook, error) {
	data, err := s.client.request(ctx, "GET", "/webhooks", nil, nil)
	if err != nil {
		return nil, err
	}

	var resp []Webhook
	if err := json.Unmarshal(data, &resp); err != nil {
		return nil, fmt.Errorf("failed to parse response: %w", err)
	}

	return resp, nil
}

// Update updates a webhook.
func (s *WebhooksService) Update(ctx context.Context, webhookID string, req *UpdateWebhookRequest) (*Webhook, error) {
	data, err := s.client.request(ctx, "PUT", "/webhooks/"+webhookID, req, nil)
	if err != nil {
		return nil, err
	}

	var resp Webhook
	if err := json.Unmarshal(data, &resp); err != nil {
		return nil, fmt.Errorf("failed to parse response: %w", err)
	}

	return &resp, nil
}

// Delete deletes a webhook.
func (s *WebhooksService) Delete(ctx context.Context, webhookID string) error {
	_, err := s.client.request(ctx, "DELETE", "/webhooks/"+webhookID, nil, nil)
	return err
}

// RotateSecret generates a new secret for a webhook.
func (s *WebhooksService) RotateSecret(ctx context.Context, webhookID string) (string, error) {
	data, err := s.client.request(ctx, "POST", "/webhooks/"+webhookID+"/rotate-secret", nil, nil)
	if err != nil {
		return "", err
	}

	var resp RotateSecretResponse
	if err := json.Unmarshal(data, &resp); err != nil {
		return "", fmt.Errorf("failed to parse response: %w", err)
	}

	return resp.Secret, nil
}

// GetCalls retrieves recent webhook delivery attempts.
func (s *WebhooksService) GetCalls(ctx context.Context, webhookID string, limit int) ([]WebhookCall, error) {
	path := fmt.Sprintf("/webhooks/%s/calls?limit=%d", webhookID, limit)
	data, err := s.client.request(ctx, "GET", path, nil, nil)
	if err != nil {
		return nil, err
	}

	var resp []WebhookCall
	if err := json.Unmarshal(data, &resp); err != nil {
		return nil, fmt.Errorf("failed to parse response: %w", err)
	}

	return resp, nil
}

// Test sends a test event to a webhook.
func (s *WebhooksService) Test(ctx context.Context, webhookID string) error {
	_, err := s.client.request(ctx, "POST", "/webhooks/"+webhookID+"/test", nil, nil)
	return err
}

File: packages/sdk-js/src/resources/webhooks.ts
import type {
  CreateWebhookRequest,
  UpdateWebhookRequest,
  Webhook,
  WebhookCall,
  RotateSecretResponse,
  ApiResponse,
} from '../types';

export class Webhooks {
  constructor(private request: <T>(method: string, path: string, body?: unknown) => Promise<T>) {}

  /**
   * Create a new webhook endpoint
   * @param data - Webhook data including URL and events to subscribe to
   */
  async create(data: CreateWebhookRequest): Promise<Webhook> {
    const response = await this.request<ApiResponse<Webhook>>(
      'POST',
      '/webhooks',
      data
    );
    return response.data;
  }

  /**
   * Get a webhook by UUID
   * @param uuid - The webhook UUID
   */
  async get(uuid: string): Promise<Webhook> {
    const response = await this.request<ApiResponse<Webhook>>(
      'GET',
      `/webhooks/${uuid}`
    );
    return response.data;
  }

  /**
   * List all webhooks for the organization
   */
  async list(): Promise<Webhook[]> {
    const response = await this.request<ApiResponse<Webhook[]>>(
      'GET',
      '/webhooks'
    );
    return response.data;
  }

  /**
   * Update a webhook
   * @param uuid - The webhook UUID
   * @param data - Fields to update
   */
  async update(uuid: string, data: UpdateWebhookRequest): Promise<Webhook> {
    const response = await this.request<ApiResponse<Webhook>>(
      'PUT',
      `/webhooks/${uuid}`,
      data
    );
    return response.data;
  }

  /**
   * Delete a webhook
   * @param uuid - The webhook UUID
   */
  async delete(uuid: string): Promise<void> {
    await this.request<ApiResponse<null>>('DELETE', `/webhooks/${uuid}`);
  }

  /**
   * Rotate the webhook secret
   * @param uuid - The webhook UUID
   */
  async rotateSecret(uuid: string): Promise<string> {
    const response = await this.request<ApiResponse<RotateSecretResponse>>(
      'POST',
      `/webhooks/${uuid}/rotate-secret`
    );
    return response.data.secret;
  }

  /**
   * Get recent webhook delivery attempts
   * @param uuid - The webhook UUID
   * @param limit - Maximum number of calls to return (default: 50)
   */
  async getCalls(uuid: string, limit?: number): Promise<WebhookCall[]> {
    const query = limit ? `?limit=${limit}` : '';
    const response = await this.request<ApiResponse<WebhookCall[]>>(
      'GET',
      `/webhooks/${uuid}/calls${query}`
    );
    return response.data;
  }

  /**
   * Send a test webhook event
   * @param uuid - The webhook UUID
   */
  async test(uuid: string): Promise<void> {
    await this.request<ApiResponse<null>>('POST', `/webhooks/${uuid}/test`);
  }
}
